_posts/2020-11-03-a-first-look-at-the-jit.md

```diff
@@ -0,0 +1,170 @@
+V8 (Javascript). These are very impressive pieces of engineering but they seem
+to have hijacked the term; not all JITs are that sophisticated, nor do they
+have to be in order to be fast.
+
+We've made many attempts at a JIT over the years that aimed for the stars only
+to fall down. Our latest and by far most successful attempt went for simplicity
+instead, trading slight inefficiencies in the generated code for ease of
+implementation. If we exclude the run-time assembler library we use, [asmjit],
+the entire thing is roughly as big as the interpreter.
+
+I believe much of our success can be attributed to four ideas we had early in
+the project:
+
+1. **All modules are always compiled to machine code.**
+
+   Previous attempts (and HiPE too) had a difficult time switching between the
+   interpreter and machine code: it was either too slow, too difficult to
+   maintain, or both.
+
+   Always running machine code means we never have to switch.
+
+2. **Data may only be kept (passed) in BEAM registers between instructions.**
+
+   This may seem silly, aren't machine registers faster?
+
+   Yes, but in practice not by much and it would make things more complicated.
+   By always passing data in BEAM registers we can use the register allocation
+   given to us by the Erlang compiler, saving us from having to do this very
+   expensive step at runtime.
+
+   More importantly, this minimizes the difference between the interpreter and
+   the JIT from the runtime system's point of view.
+
+3. **Modules are compiled one instruction at a time.**
+
+   One of the most difficult problems in our prior attempts was to strike a
+   good balance between the time it took to compile something and the eagerness
+   to do so. If we're too eager, we'll spend too much time compiling, and if
+   we're too lax we won't see any improvements.
+
+   This problem was largely self-inflicted and caused by the compiler being too
+   slow (we often used LLVM), which was made worse by us giving it large pieces
+   of code to allow more optimizations.
+
+   By limiting ourselves to compiling one instruction at a time, we leave some
+   performance on the table but greatly improve compilation speed.
+
+4. **Every instruction has a handwritten machine code template.**
+
+   This makes compilation _extremely_ fast as we basically just copy-paste
+   the template every time the instruction is used, only performing some minor
+   tweaks depending on its arguments.
+
+   This may seem daunting at first but it's actually not that bad once you get
+   used to it. While it certainly takes a lot of code to achieve even the
+   smallest of things, it's inherently simple and easy to follow as long as
+   the code is kept short.
+
+   The downside is that every instruction needs to be implemented for each
+   architecture, but luckily there's not a lot of popular ones and we plan to
+   support the two most common ones by the time we release OTP 24; `x64` and
```

Comment:
```suggestion
   support the two most common ones by the time we release OTP 24: `x64` and
```

Reply:
I could've sworn I fixed that before posting, thanks for pointing it out!